017 - Storage Buckets (#25)
appwrite/rfc: 017-storage-buckets.md

-->
You can use buckets to organize your storage and control access to your files, but unlike directories and folders (which are not part of the Storage API anyway), you cannot nest buckets.

In the Storage Service, Buckets are going to act like a folder that will add logic to how the API responds. A Bucket has following configurations available:
- **Enabled**
  - This setting decides if a Bucket is enabled and is accessible.
- **Adapter**
  - This decides to what Storage Adapter the files will be uploaded to. Possible options are Local, S3, DigitaOcean Spaces, etc. This is not allowed to be changed after a file has been added to the bucket.
- **Maximum File Size Limit**
  - This setting decides the maximum single file size that can be uploaded.
- **Maximum Number of Files**
  - This setting decides how many files a Bucket can contain.
- **Accepted File Extensions**
  - This setting decides what file extensions can be uploaded.
- **Encryption**
  - This setting decides if files are encrypted. This will come into play, when encrypting large files causes too much load. This is not allowed to be changed after a file has been added to the bucket.
- **Virus Scan**

>   - This setting decides if files are scanned by ClamAV after upload. Same reasoning ass **Encryption**.

Reviewer: ```suggestion
  - This setting decides if files are scanned by ClamAV after upload. Same reasoning as **Encryption**.
```

> - **TTL**

Reviewer 2: We should make sure we calculate storage usage properly when TTL expires

Reviewer 2: Should we also allow TTL per file?

Author: Added 👍🏻

  - This setting decides how long a file is supposed to alive. Files are going to be deleted automatically. This is not allowed to be changed after a file has been added to the bucket.
- **Permission**
  - This setting decides what permission role is allowed to upload files to the bucket. Read and Write permissions will still exist for each file and be the source of permission check.

A Bucket will have following Model in our internal Database:
```typescript
{
  $id: string;
  $write: string[];
  name: string;
  enabled: boolean;
  adapter: 'local'|'s3'|'etc'
  encrypted: boolean;
  antivirus: boolean;
  adapterConfig?: object;
  ttl?: number;
  maximumFiles?: number;
  maximumFileSize?: number;
  allowedFileExtensions?: string;
}
```

### New Endpoints

In this section I will explain all necessary endpoints for integratting Storge Buckets.

> #### `GET: /v1/storage/bucket/`

Reviewer 2: In order to be persistent with our current conventions this should be `/v1/storage/buckets/`

This endpoint lists all Buckets. *Server*

Payload:
- **search = ''** Search term to filter your list results.
- **limit = 25** Maximum number of Buckets to return.
- **offset = 0** Offset value.

#### `GET: /v1/storage/bucket/{bucketId}`

This endpoint returns a single Bucket by its unique ID. *Server*

#### `GET: /v1/storage/bucket/{bucketId}/files`

This endpoint returns a list of all Files in a specific Bucket. *Server & Client*

Payload:
- **filters = []** Array of filter strings.
- **limit = 25** Maximum number of Buckets to return.
- **offset = 0** Offset value.
- **search = ''** Search query.

> #### `POST: /v1/storage/bucket`

Reviewer 2: To simplify the API, and avoid passing complex object I think we should seriously consider have an individual POST endpoint for each adapter. Each endpoint will be able to receive primitive params that the adapter needs. Internally we can store the data as an object. A unique index for a unique adapter will make the documentation and API easier to understand and will be more consistent with current behavior that favorite the usage of plain primitive params.

For example:
``POST: /v1/storage/bucket/s3`
basic params for bucket creation + 
string - accessKey
string- secretKey
string - bucketName
string - region
string - aclFlag

Author: Added.

This endpoint creates a new Bucket. *Server*

Payload:
- **name** Bucket name.
- **write** An array of strings with write permissions.
- **encrypted = true** Enable Encryption.
- **antivirus = true** Enable Anti Virus.
- **adapter = 'local'** Adapter to use.
- **adapterConfig = null** Adapter Config.
- **ttl = null** TTL for files.
- **maximumFiles = null** Maximum amount of files.
- **maximumFileSize = null** Maximum file size.
- **allowedFileExtensions = null** Allowed File extensions.

#### `PUT: /v1/storage/bucket/{bucketId}`

This endpoint updates a Bucket by its unique ID. *Server*

Payload:
- **name = null** Bucket name.
- **write = null** An array of strings with write permissions.
- **encrypted = null** Enable Encryption.
- **antivirus = null** Enable Anti Virus.
- **adapter = null** Adapter to use.
- **adapterConfig = null** Adapter Config.
- **ttl = null** TTL for files.

> - **maximumFiles = null** Maximum amount of files.

Reviewer 2: What happens when we are over capacity after update?

Author: Same with the user limits - we won't allow any more uploads.

- **maximumFileSize = null** Maximum file size.
- **allowedFileExtensions = null** Allowed File extensions.

#### `DELETE: /v1/storage/bucket/{bucketId}`

> This endpoint deletes a Bucket by its unique ID. *Server*

Reviewer 2: This endpoint should trigger a background delete of all files in the bucket from both db and storage. This action should probably be delayed by x time to avoid abuse or actions by unauthorized users. The delay will give a time window to revert the action on the server admin level.

Reviewer 2: We might want to do something similar on a collection delete

Author: Added.

Reviewer 2: We should also add and emphasis the **delayed** part, if a malicious user is triggering a full bucket delete we want to have a timeframe where we can still stop the operation.

  you have some relevant papers to refer to, this can serve as a more detailed
  theoretical background.

This section is intended to encourage you as an author to think about the
lessons from other software, provide readers of your RFC with a fuller picture.
If there is no prior art, that is fine - your ideas are interesting to us
whether they are brand new or if it is an adaptation from other software.

Write your answer below.
-->
- [About Storage Buckets](https://docs.uipath.com/orchestrator/docs/about-storage-buckets)
- [Google Cloud Storage](https://cloud.google.com/storage/docs/json_api/v1/buckets)
- [Google Cloud Storage - Key Terms](https://cloud.google.com/storage/docs/key-terms)

### Unresolved questions

[unresolved-questions]: #unresolved-questions

<!-- What parts of the design do you expect to resolve through the RFC process before this gets merged? -->

#### Prevent Breaking Changes

To prevent a huge breaking change in the Storage Services API's - we are allowing users to use the current Storage implementation as well. Now there is the problem of offering a consistent API pattern across the services.

> Let's take a look at the `GET: /v1/storage/files/{fileId}` endpoint. This will result into retrieving a file from the current Storage service. Now if a file is inside a Bucket - the endpoint would have to look like `GET: /v1/storage/files/{bucketId}/{fileId}`. This becomes problematic, since right now it is not possible to allow optional Parameters in the Path with the same Controller/SDK Method and would require refactoring in all SDKs, Utopia and Appwrite.

Reviewer: 1. I like the endpoint `/v1/storage/files/{bucketId}/{fileId}
2. If no bucket is created may be appwrite can handle the bucket as `default` internally
3. Internally any path passed only as `/v1/storage/files/{fileId}` appwrite internally converts to `/v1/storage/files/default/{fileId}`

Reviewer 2: We will do something like that **temporarily** for backward compatibility, but this will not be consisted with the way we handle all other endpoint paths across Appwrite ([service]/[entity]/[id]).

Possible solutions can be the following:

- Add the targeted Bucket as a query param to `GET: /v1/storage/files/default/{fileId}?bucket={bucketId}`.
  - This breaks our API Standards - but could be a good tradeoff in terms of the migration for develoeprs, since there will be no breaking change.
- Adding a keyword passing to the Bucket parameter that falls back to the default Storage `GET: /v1/storage/files/default/{fileId}`.
  - This can be done without introducing any breaking changes to the SDK usage
- Allowing to do calls to `GET: /v1/storage/files/YYYY/XXXX` and `GET: /v1/storage/files//XXXX`
  - Not a 100% sure this is compliant with any REST Standards

> - Having 2 separate endpoints for each, the default Storage and Storage Buckets, with its own SDK methods

Reviewer 2: `Having 2 separate endpoints for each, the default Storage and Storage Buckets, with its own SDK methods`

I think this is the best and simplest option for us. We can change the names of old endpoints in the new SDK to have a prefix we decide.

  name: string;
  enabled: boolean;
  adapter: 'local'|'s3'|'etc'
  encrypted: boolean;
  antivirus: boolean;
  ttl?: number;
  maximumFiles?: number;
  maximumFileSize?: number;
  allowedFileExtensions?: string;
}
```

In terms of a storage adapter except `Local`, additional values will be present like configurations for S3.

### New Endpoints

In this section I will explain all necessary endpoints for integratting Storge Buckets.

#### `GET: /v1/storage/buckets/`

This endpoint lists all Buckets. *Server*

Payload:
- **search = ''** Search term to filter your list results.
- **limit = 25** Maximum number of Buckets to return.
- **offset = 0** Offset value.

#### `GET: /v1/storage/buckets/{bucketId}`

This endpoint returns a single Bucket by its unique ID. *Server*

#### `GET: /v1/storage/buckets/{bucketId}/files`

This endpoint returns a list of all Files in a specific Bucket. *Server & Client*

Payload:
- **filters = []** Array of filter strings.
- **limit = 25** Maximum number of Buckets to return.
- **offset = 0** Offset value.
- **search = ''** Search query.

#### `POST: /v1/storage/buckets/{adapter}`

This endpoint creates a new Bucket. *Server*

Payload:

> - **name** Bucket name.
> - **write** An array of strings with write permissions.
> - **encrypted = true** Enable Encryption.
> - **antivirus = true** Enable Anti Virus.
> - **ttl = false** Enables TTL for files.
> - **maximumFiles = null** Maximum amount of files.
> - **maximumFileSize = null** Maximum file size.
> - **allowedFileExtensions = null** Allowed File extensions.

Reviewer 2: ```suggestion
- **name** Bucket name.
- **write** An array of strings with write permissions.
- **encrypted = true** Enable Encryption.
- **antivirus = true** Enable Anti Virus.
- **ttl = false** Enables TTL for files.
- **maxFiles = null** Maximum amount of files.
- **maxFileSize = null** Maximum file size.
- **fileExtensions = null** Allowed File extensions. Leave empty to allow any extension.
```

Added a few suggestions to param names - still not 100% sure about them.

I think we should consider to implement bot `maxFiles` and `ttl` options in a later stage, not sure how well the DB will scale with them.

Additional Payload for **Local** (`/v1/storage/buckets/local`):
- *none*

Additional Payload for **S3** (`/v1/storage/buckets/s3`):

> - **adapterAccessKeyId** S3 Access Key ID.
> - **adapterSecretKey** S3 Secret Key.
> - **adapterBucket** S3 Bucket Name.
> - **adapterRegion** S3 Region.
> - **adapterAcl** S3 Access Control List.

Reviewer 2: ```suggestion
- **accessKeyId** S3 Access Key ID.
- **secretKey** S3 Secret Key.
- **bucket** S3 Bucket Name.
- **region** S3 Region.
- **acl** S3 Access Control List.
```

The adapter context is already clear from the path/title
